src/server/worker/api_server.go

```diff
@@ -1600,42 +1730,135 @@ func (a *APIServer) worker() {
 	})
 }
 
+func (a *APIServer) claimShard(ctx context.Context) {
+	shards := a.shards.ReadOnly(ctx)
+	watcher, err := shards.Watch()
+	if err != nil {
+		log.Printf("error creating shard watcher: %v", err)
+		return
+	}
+	// attempt to claim a shard as they become available
+	claimedShard := noShard
+CLAIMED_SHARD:
+	for {
+		for shard := int64(0); shard < a.numShards; shard++ {
+			var shardInfo ShardInfo
+			if err := shards.Get(fmt.Sprint(shard), &shardInfo); err != nil {
+				if !col.IsErrNotFound(err) {
+					log.Printf("error checking shards: %v", err)
+					return
+				}
+				// attempt to claim shard
+				if _, err := col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
+					shards := a.shards.ReadWrite(stm)
+					var shardInfo ShardInfo
+					if err := shards.Get(fmt.Sprint(shard), &shardInfo); err != nil {
+						if !col.IsErrNotFound(err) {
+							return err
+						}
+						claimedShard = shard
+						return shards.PutTTL(fmt.Sprint(shard), &ShardInfo{}, shardTTL)
+					}
+					claimedShard = noShard
+					return nil
+				}); err != nil {
+					log.Printf("shard claim error: %v", err)
+					return
+				}
+				if claimedShard != noShard {
+					break CLAIMED_SHARD
+				}
+			}
+		}
+		// wait for a deletion event (ttl expired) to attempt to acquire a shard again
+		var e *watch.Event
+		for e == nil || e.Type != watch.EventDelete {
+			select {
+			case e = <-watcher.Watch():
+				if e.Type == watch.EventError {
+					log.Printf("shard watch error: %v", e.Err)
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}
+	a.clients = make(map[string]Client)
+	a.claimedShard <- claimedShard
+	// renew lock on shard
+	for {
+		select {
+		case <-time.After((time.Second * time.Duration(shardTTL)) / 2):
+			if _, err := col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
+				shards := a.shards.ReadWrite(stm)
+				var shardInfo ShardInfo
+				if err := shards.Get(fmt.Sprint(claimedShard), &shardInfo); err != nil {
+					return err
+				}
+				return shards.PutTTL(fmt.Sprint(claimedShard), &shardInfo, shardTTL)
+			}); err != nil {
+				log.Printf("failed to renew lock on shard: %v", err)
```

Comment:
So, I decided to go with a simple channel that communicates that a shard was lost because I think communicating this through a context might not be a good design pattern considering the code flow.

Reply:
I don't think what you did quite fixes the problem, because it only checks `lostShard` at the beginning of `mergeDatums` but you can still lose the shard after you're past that point. I think this is a case where you kind of need `ctx`, because that's the only way to cancel an ongoing rpc which you need to be able to do when the shard gets lost.
Also I think the code as is will deadlock as is because `lostShard` isn't initialized (unless I'm missing it). `nil` channel is valid in go but is channel that blocks all sends and receives. I think you'd run a risk of deadlocks even if it were initialized because you send to the channel in a loop in `claimShard` so eventually you'll fill it up and then that goro will be blocked and you won't be able to claim shards anymore.

These are the types of edge cases I always worry about with channels and why I prefer to avoid them in cases like this where I think passing a closure works a lot more cleanly. Another similar case is function based iteration vs a chan based iterator type.

Reply:
Okay, so there are a few things going on here. The first thing I am going to address is the uninitialized lostShard channel, that is clearly a stupid mistake and should be initialized to a buffered channel of size 1. Without it being initialized, the deadlock situation would be that a worker would never lose a shard due to taking too long to renew (which is only a deadlock in the sense that claim shard is trying to write to the channel, this would not impact jobs completing). I do not think there is a deadlock situation when it is initialized though because after failing to renew, we return from the claimShard function. The channel will only get written to once. But, I think it is probably going to make sense to have the worker capable of claiming a shard in the future even if it failed to renew a shard at some point, so there probably are still some changes needed.

As far as context vs channel, I still think I am leaning towards a channel simply because there does not seem to be a clear way to incorporate a context for a shard into the context hierarchy we have for the job. Since we want to have workers retain shards across jobs (to avoid re-downloading hashtrees and potentially data) we would need a context that is out-of-band with the context that we use for jobs because cancellation of the job context should not correspond to a lost shard. It seems like the only way to get this behavior would be to derive it from the job context or some other child context (which really doesn't make sense because that claim has nothing to do with a particular job) or create a separate context. If we go the separate context route, then what purpose does it serve other than to use the done channel and we would not even be able to use it in the APIs we use because they take one context and we are going to preference the job context over some shard context in that case. The merge datums operation is idempotent, so if in some cases, which I believe should be rare, we do some duplicate merging I think we should be fine. Unless there is some way to use contexts that I am unaware of that will address the issues I mentioned, then I think a simple channel is the most straightforward solution.